vmm/src/device_manager.rs

```diff
@@ -1161,6 +1163,41 @@ impl DeviceManager {
     pub fn get_gic_device_entity(&self) -> Option<&Arc<Mutex<Box<dyn GICDevice>>>> {
         self.gic_device_entity.as_ref()
     }
+    #[cfg(target_arch = "aarch64")]
+    pub fn construct_gicr_typers(&self, vcpu_states: &[CpuState]) {
```

Comment:
Hi @sboeuf. Thanks very much for the comment! Actually I hope I would have a discussion about the design about the GIC before I do any modifications, as I am also not sure current code design makes very good sense:

My original thought would be: In current design for AArch64, the software-implemented [devices/src/gic](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/devices/src/gic.rs) is only to provide interrupt delivery service, and the real vGIC device is created through the ioctl here: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/5a3d54fa1945b8c45122e1af33382c7337d869f0/arch/src/aarch64/mod.rs#L153

When we need to save VM, we need to save/restore the vGIC, and I think based on the current design of the VM structure https://github.com/cloud-hypervisor/cloud-hypervisor/blob/5a3d54fa1945b8c45122e1af33382c7337d869f0/vmm/src/vm.rs#L288 we can only access the real vGIC device through the device_manager. This is the reason why the `gic_device_entity` field and related methods (`get/set_gic_device_entity`, `construct_gicr_typers`) were introduced in commit 8e0713a2810f7339afd50b928c99343ecac294e1.

I did think of implement the `Snapshottable` trait for GIC in [devices/src/gic](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/devices/src/gic.rs). However later I noticed that for GICv2, GICv3 and GICv3ITS, the registers (or states) that need to be saved are different. So I think the `Snapshottable` trait should be implemented separately in arch/src/gic. As the real vGIC device is created using trait object, we need to downcast the trait object to specific type. 

Hopefully now things is more clear, any suggestions about how we can optimise the design?

Reply:
Oh I understand now. Well then if only the device manager can access it, and also because this is a common piece of code for all GIC devices, let's keep it where it is :)
Thanks for the explanation!